ArduPlane/quadplane.cpp

```diff
@@ -2006,15 +2079,34 @@ void QuadPlane::vtol_position_controller(void)
 
     case QPOS_LAND_DESCEND: {
         float height_above_ground = plane.relative_ground_altitude(plane.g.rangefinder_landing);
-        pos_control->set_alt_target_from_climb_rate(-landing_descent_rate_cms(height_above_ground),
-                                                    plane.G_Dt, true);
+        if (doing_precision_landing && plane.rangefinder_alt_ok() && height_above_ground > 35.0f && height_above_ground < 200.0f) {
+            float max_descent_speed = abs(land_speed_cms)/2.0f;
+            float land_slowdown = MAX(0.0f, pos_control->get_horizontal_error()*(max_descent_speed/precland_acceptable_error));
+            cmb_rate = MIN(-precland_min_descent_speed, -max_descent_speed+land_slowdown);
+            pos_control->set_alt_target_from_climb_rate_ff(cmb_rate, plane.G_Dt, true);
+        }
+        else
+        {
+            pos_control->set_alt_target_from_climb_rate(-landing_descent_rate_cms(height_above_ground),
+                                                         plane.G_Dt, true);
+        }
         break;
     }
 
-    case QPOS_LAND_FINAL:
-        pos_control->set_alt_target_from_climb_rate(-land_speed_cms, plane.G_Dt, true);
+    case QPOS_LAND_FINAL: {
+        float height_above_ground = plane.relative_ground_altitude(plane.g.rangefinder_landing);
```

Comment:
This feels like it could be factored.

Reply:
I agree with you. Still thinking about the best way to do this. The best way I have so far was to create a helper function that returns land_speed_cms but I wasn't sure if that was the best way to go. Still thinking about this one.

Reply:
I cleaned this up a little. Let me  know what you think.